### PR TITLE
fix: Ensure importlib_metadata runtime dependency installed with Python < 3.10

### PR DESCRIPTION
**Pull request by jcfr**
Fixes a regression introduced in 5f068d5 ("chore(build): move to scikitbuild-core (#455)", 2024-03-01)

### CODE REVIEW

**Comment by jcfr** [comment]
cc: @mayeut @henryiii

**Comment by jcfr** [comment]
Thanks @agriyakhetarpal, @braindevices, @mloubout and @roastduck for the issue reports and pull requests :pray:

**Comment by jcfr** [comment]
Once the fix has been reviewed, approved and integrated, suggested path forward:
- Yank CMake `3.29.0`
- Release `3.29.0.1`

**Comment by mloubout** [comment]
Still needs to fix from #473  for correct version checking

**Comment by agriyakhetarpal** [comment]
@mudit2812 and @mloubout, Python 3.7 support was recently dropped with `3.29.0`. Do we even need the conditional import for `importlib_metadata` at all, then – We can just use `importlib.metadata` and that should work?

**Comment by mloubout** [comment]
Didn't see Python 3.8 was the minimum version now, should be able to completely drop `importlib_metadata` then as you pointed out.

File: pyproject.toml [hunk 1]
```diff
@@ -24,6 +24,9 @@ classifiers = [
     "Topic :: Software Development :: Build Tools",
     "Typing :: Typed"
 ]
+dependencies = [
+  "importlib_metadata>=2.0; python_version<'3.10'",
+]
```
Comment (agriyakhetarpal) [inline pyproject.toml:29]:
```suggestion
```

should not be needed – please see https://github.com/scikit-build/cmake-python-distributions/pull/474#issuecomment-2020766395 :)

**Comment by agriyakhetarpal** [comment]
`docs/conf.py` seems to have been using the correct Python version check, as reported by @mloubout:

https://github.com/scikit-build/cmake-python-distributions/blob/85a457bf3e0de6efa65cc0f47f017810936772f0/docs/conf.py#L20-L25

**Comment by henryiii** [comment]
Python 3.7 is the minimum version, not 3.8. We just dropped 2.7 and 3.6.

**Comment by agriyakhetarpal** [comment]
Oops, thanks! Updated my review to mark that change

File: src/cmake/__init__.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 import subprocess
 import sys
 
-if sys.version_info < (3, 10):
+if sys.version_info < (3, 8):
     from importlib_metadata import distribution
 else:
     from importlib.metadata import distribution
```
Comment (jcfr) [inline src/cmake/__init__.py:8]: I don't think `importlib.metadata.distribution` is available with Python < 3.10

| Python Version  | `importlib.metadata` | URL |
| --|--|--|
| 3.8  | :x: | https://github.com/python/cpython/tree/3.8/Lib/importlib |
| 3.9  | :x: | https://github.com/python/cpython/tree/3.9/Lib/importlib |
| 3.10 | :white_check_mark:  | https://github.com/python/cpython/tree/3.10/Lib/importlib|

Comment (jcfr) [reply src/cmake/__init__.py:8]: I then suggest to keep the original patch

**Comment by henryiii** [comment]
@jcfr I don't have permission to yank `cmake 3.29.0`, let's go ahead and do that, since it's breaking people and people probably aren't requiring it yet after a few hours, and if they did, it would still work (due to yanks resolving if they can't resolve to a non-yanked version). CI will take ~20-30 mins here, then the release will take ~3 hours.

File: src/cmake/__init__.py (hunk 2 above)
Comment (henryiii) [reply src/cmake/__init__.py:8]: * https://github.com/python/cpython/blob/95c340a86b828cac6c0a2e4f2fa8a96695388c73/Lib/importlib/metadata.py#L27
* https://github.com/python/cpython/blob/95c340a86b828cac6c0a2e4f2fa8a96695388c73/Lib/importlib/metadata.py#L27

?

Comment (mloubout) [reply src/cmake/__init__.py:8]: It is:

```
^[[Amathias@wampa:~$ docker run --rm -it python:3.8 python3 -c 'from importlib_metadata import distribution'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'importlib_metadata'
mathias@wampa:~$ docker run --rm -it python:3.8 python3 -c 'from importlib.metadata import distribution'
```

Comment (mloubout) [reply src/cmake/__init__.py:8]: https://github.com/python/cpython/blob/95c340a86b828cac6c0a2e4f2fa8a96695388c73/Lib/importlib/metadata.py#L497

Comment (henryiii) [reply src/cmake/__init__.py:8]: (I manually checked this on 3.8 before https://github.com/scikit-build/cmake-python-distributions/pull/474#issuecomment-2020805647 after checking the docs)

```PyCon
$ python3.8                                                                                                                        
>>> import importlib.metadata
>>> importlib.metadata.distribution("pip").files[0].locate().resolve()
PosixPath('/usr/local/lib/python3.8/site-packages/AUTHORS.txt')
```

**Comment by jcfr** [comment]
> re: yanked

image

File: src/cmake/__init__.py (hunk 2 above)
Comment (jcfr) [reply src/cmake/__init__.py:8]: Thanks for clarifying, not seeing the "metadata" directory tricked me.

**Comment by jcfr** [comment]
Thanks @henryiii for further tweaking the patch :pray:  and @mloubout for testing.

:rocket: :heavy_check_mark:

**Review by mloubout — APPROVED** [review]
Great, thanks for the quick resolution 🔥

**Pull request merged**